spmd/tensor/ops/tensor_ops.py

```diff
@@ -111,3 +120,161 @@ def prop_bucketize(op_schema: OpSchema) -> OutputSharding:
+        suggested_input_spec = DTensorSpec(
+            mesh=input_spec.mesh,
+            placements=output_placements,
+            ndim=input_spec.ndim,
+            shape=input_spec.shape,
+        )
+        out = OutputSharding(
+            output_spec=None,
+            schema_suggestions=[
+                OpSchema(
+                    args_schema=(suggested_input_spec,)
+                    + op_schema.args_schema[1:],
+                    kwargs_schema=op_schema.kwargs_schema,
+                ),
+            ],
+        )
+    return out
+
+
+@register_prop_rule("aten.slice.Tensor")
+def prop_slice(op_schema: OpSchema) -> OutputSharding:
+    """NOTE: can be further optimized (right now it replicates before slicing on a sharded dimension)"""
+    defaults = (None, 0, None, None, 1)
+    input_spec, dim, start, end, step = (
+        op_schema.args_schema + defaults[len(op_schema.args_schema) :]
+    )
+    assert isinstance(input_spec, DTensorSpec)
+    assert isinstance(dim, int)
+    assert start is None or isinstance(start, int)
+    assert end is None or isinstance(end, int)
+    assert isinstance(step, int)
+
+    # normalize arguments
+    if dim < 0:
+        dim += input_spec.ndim
+    if start is None:
+        start = 0
+    if step is None:
+        step = 1
+    if end is None or end > input_spec.shape[dim]:
+        end = input_spec.shape[dim]
+    if end < 0:
+        end += input_spec.shape[dim]
+
+    if start == 0 and end == input_spec.shape[dim] and step == 1:
+        return OutputSharding(output_spec=input_spec)
+
+    # shape propagation
+    slice_len = (end - start + step - 1) // step
+    out_shape = torch.Size(
+        tuple(input_spec.shape[0:dim])
+        + (slice_len,)
+        + tuple(input_spec.shape[dim + 1 :])
+    )
+
+    return _prop_all_but_dim(op_schema, dim=dim, out_shape=out_shape)
+
+
+@register_prop_rule("aten.slice_scatter.default")
+def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+    # 1. number of dimensions in input and src need to match.
+    # 2. number of elements on all non-dim need to match between input and src.
+    # 3. numer of elements in src in dim need to match the slice size.
+    # Given the above:
+    # - We suggest for src to follow the sharding of input, except on the scatter dimension,
+    #   where our best bet for now is to make them replicated as a fall-back.
+    #   TODO: Ideally we'd like to make sure the output is re-sharded afterwards to keep input sharding.
+
+    defaults = (None, None, 0, None, None, 1)
+    input, src, dim, start, end, step = (
+        op_schema.args_schema + defaults[len(op_schema.args_schema) :]
+    )
+    assert isinstance(input, DTensorSpec)
+    assert isinstance(src, DTensorSpec)
+    assert isinstance(dim, int)
+
+    if dim < 0:
+        dim += input.ndim
+
+    # first, we keep the input sharding, except for the input dimension
+    # also, we cannot allow partial sum anymore.
+    input_suggestion = tuple(
```

Comment:
can we merge this with `unshard_tensor_dim`? or is it because `aten.slice.Tensor` support forwarding partial?

Reply:
right, looking back i don't like unshard_tensor_dim since it's just a single line of code; i tought of adding allow_partial=True but then it would make it harder to read.